File: src/test/java/io/github/numichi/reactive/logger/reactor/MDCContextTest.java
package io.github.numichi.reactive.logger.reactor;

import io.github.numichi.reactive.logger.MDC;
import io.github.numichi.reactive.logger.exception.InvalidContextDataException;
import io.github.numichi.reactive.logger.java.MDCContext;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import reactor.core.publisher.Mono;
import reactor.test.StepVerifier;
import reactor.util.context.Context;

import static io.github.numichi.reactive.logger.Values.DEFAULT_REACTOR_CONTEXT_MDC_KEY;
import static org.junit.jupiter.api.Assertions.assertThrows;

class MDCContextTest {
    
    public static final String ANOTHER_CONTEXT_KEY = "another-context-key";
    
    @Test
    @DisplayName("should give the MDC you are looking for")
    void shouldGiveTheMDCByDefault() {
        MDC mdc1 = new MDC();
        mdc1.put("mdcKey", "mdcValue");
        
        MDC mdc2 = new MDC(ANOTHER_CONTEXT_KEY);
        mdc2.put("mdcKey", "mdcValue");
        
        
        Mono<MDC> resultDefaultByContext = Mono.deferContextual(MDCContext::read)
            .contextWrite(it -> MDCContext.put(it, mdc1));
        
        StepVerifier.create(resultDefaultByContext)
            .expectNext(mdc1)
            .verifyComplete();
        
        
        Mono<MDC> resultDefault = Mono.defer(MDCContext::read)
            .contextWrite(it -> MDCContext.put(it, mdc1));
        
        StepVerifier.create(resultDefault)
            .expectNext(mdc1)
            .verifyComplete();
        
        
        Mono<MDC> resultAnotherByContext = Mono.deferContextual(ctx -> MDCContext.read(ctx, ANOTHER_CONTEXT_KEY))
            .contextWrite(it -> MDCContext.put(it, mdc2));
        
        StepVerifier.create(resultAnotherByContext)
            .expectNext(mdc2)
            .verifyComplete();
        
        
        Mono<MDC> resultAnother = Mono.defer(() -> MDCContext.read(ANOTHER_CONTEXT_KEY))
            .contextWrite(it -> MDCContext.put(it, mdc2));
        
        StepVerifier.create(resultAnother)
            .expectNext(mdc2)
            .verifyComplete();
    }
    
    @Test
    @DisplayName("should return the appropriate contexts count")
    void shouldReturnTheAppropriateContextsCount() {
        MDC defaultMdc = new MDC();
        defaultMdc.put("mdcKey", "mdcValue");
    
        MDC anotherMdc = new MDC(ANOTHER_CONTEXT_KEY);
        anotherMdc.put("mdcKey", "mdcValue");
    
        Mono<Integer> contextSize1 = Mono.deferContextual(ctx -> Mono.just(ctx.size()))
            .contextWrite(it -> MDCContext.put(it, defaultMdc))
            .contextWrite(it -> MDCContext.put(it, anotherMdc));
        StepVerifier.create(contextSize1)
            .expectNext(2)
            .verifyComplete();
    
    
        Mono<Integer> contextSize2 = Mono.deferContextual(ctx -> Mono.just(ctx.size()))
            .contextWrite(it -> MDCContext.put(it, defaultMdc, anotherMdc));
        StepVerifier.create(contextSize2)
            .expectNext(2)
            .verifyComplete();
    
    
        Mono<Integer> contextSize3 = Mono.deferContextual(ctx -> Mono.just(ctx.size()))
            .contextWrite(it -> MDCContext.put(it, defaultMdc, null));
        StepVerifier.create(contextSize3)
            .expectNext(1)
            .verifyComplete();
    
        
        Mono<Integer> contextSize4 = Mono.deferContextual(ctx -> Mono.just(ctx.size()))
            .contextWrite(it -> MDCContext.put(it, defaultMdc, null))
            .contextWrite(it -> it.put("A", "B"));
        StepVerifier.create(contextSize4)
            .expectNext(2)
            .verifyComplete();
    }
    
    @Test
    @DisplayName("should give the MDC you are looking for (more MDC Context)")
    void shouldGiveTheMDCWithMoreMdcContext() {
        MDC defaultMdc = new MDC();
        defaultMdc.put("mdcKey", "mdcValue");
        
        MDC anotherMdc = new MDC(ANOTHER_CONTEXT_KEY);
        anotherMdc.put("mdcKey", "mdcValue");
        
        
        Mono<MDC> resultDefault = Mono.defer(MDCContext::read)
            .contextWrite(it -> MDCContext.put(it, defaultMdc))
            .contextWrite(it -> MDCContext.put(it, anotherMdc));
        StepVerifier.create(resultDefault)
            .expectNext(defaultMdc)
            .verifyComplete();
        
        
        Mono<MDC> resultAnother = Mono.defer(() -> MDCContext.read(ANOTHER_CONTEXT_KEY))
            .contextWrite(it -> MDCContext.put(it, defaultMdc))
            .contextWrite(it -> MDCContext.put(it, anotherMdc));
        StepVerifier.create(resultAnother)
            .expectNext(anotherMdc)
            .verifyComplete();
    }
    
    @Test
    @DisplayName("should throw IllegalArgumentException if any parameter is NULL")
    void shouldThrowIllegalArgumentExceptionTest() {
        assertThrows(IllegalArgumentException.class, () -> MDCContext.put(null, new MDC()));
        assertThrows(IllegalArgumentException.class, () -> MDCContext.put(Context.empty(), null));
    
        StepVerifier.create(MDCContext.read((String) null))
            .expectError(IllegalArgumentException.class)
            .verify();
    
        StepVerifier.create(MDCContext.read(null, ""))
            .expectError(IllegalArgumentException.class)
            .verify();
    
        StepVerifier.create(MDCContext.read(Context.of("", ""), null))
            .expectError(IllegalArgumentException.class)
            .verify();
    }
    
    @Test
    @DisplayName("should throw InvalidContextDataException if context data is invalid")
    void shouldThrowInvalidContextDataException() {
        // ANOTHER_CONTEXT_KEY is not DEFAULT_REACTOR_CONTEXT_MDC_KEY
        MDC mdc = new MDC(ANOTHER_CONTEXT_KEY);
        mdc.put("mdcKey2", "mdcValue2");
        Mono<MDC> result1 = Mono.defer(MDCContext::read)
            .contextWrite(it -> MDCContext.put(it, mdc));
        StepVerifier.create(result1)
            .expectError(InvalidContextDataException.class)
            .verify();
    
        
        Mono<MDC> result2 = Mono.defer(() -> MDCContext.read("not-exist-context-id"))
            .contextWrite(it -> MDCContext.put(it, mdc));
        StepVerifier.create(result2)
            .expectError(InvalidContextDataException.class)
            .verify();
        
        
        Mono<MDC> result3 = Mono.defer(MDCContext::read)
            .contextWrite(it -> it.put(DEFAULT_REACTOR_CONTEXT_MDC_KEY, ""));
        StepVerifier.create(result3)
            .expectError(InvalidContextDataException.class)
            .verify();
        
        
        Mono<MDC> result4 = Mono.defer(MDCContext::read)
            .contextWrite(it -> it.put(DEFAULT_REACTOR_CONTEXT_MDC_KEY, 100));
        StepVerifier.create(result4)
            .expectError(InvalidContextDataException.class)
            .verify();
    }
}

File: src/test/java/io/github/numichi/reactive/logger/reactor/MDCContextTest.kt
package io.github.numichi.reactive.logger.reactor

import io.github.numichi.reactive.logger.Configuration
import io.github.numichi.reactive.logger.DEFAULT_REACTOR_CONTEXT_MDC_KEY
import io.github.numichi.reactive.logger.MDC
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import reactor.core.publisher.Mono
import reactor.test.StepVerifier

internal class MDCContextTest {

    @BeforeEach
    fun setUp() {
        Configuration.reset()
    }

    @Test
    fun `should give the MDC you are looking for`() {
        run {
            val mdc = MDC()
            mdc["mdcKey"] = "mdcValue"

            val resultDefaultByContext: Mono<MDC> = Mono.deferContextual { MDCContext.read(it) }
                .contextWrite { MDCContext.put(it, mdc) }

            StepVerifier.create(resultDefaultByContext)
                .expectNext(mdc)
                .verifyComplete()

            val resultDefault: Mono<MDC> = MDCContext.read()
                .contextWrite { MDCContext.put(it, mdc) }

            StepVerifier.create(resultDefault)
                .expectNext(mdc)
                .verifyComplete()
        }

        run {
            val mdc = MDC(ANOTHER_CONTEXT_KEY)
            mdc["mdcKey"] = "mdcValue"

            val resultAnotherByContext: Mono<MDC> = Mono.deferContextual { MDCContext.read(it, ANOTHER_CONTEXT_KEY) }
                .contextWrite { MDCContext.put(it, mdc) }

            StepVerifier.create(resultAnotherByContext)
                .expectNext(mdc)
                .verifyComplete()

            val resultAnother: Mono<MDC> = MDCContext.read(ANOTHER_CONTEXT_KEY)
                .contextWrite { MDCContext.put(it, mdc) }

            StepVerifier.create(resultAnother)
                .expectNext(mdc)
                .verifyComplete()
        }
    }

    @Test
    fun `should return the appropriate contexts count`() {
        val defaultMdc = MDC()
        defaultMdc["mdcKey"] = "mdcValue"

        val anotherMdc = MDC(ANOTHER_CONTEXT_KEY)
        anotherMdc["mdcKey"] = "mdcValue"

        run {
            val mono = Mono.deferContextual { Mono.just(it.size()) }
                .contextWrite { MDCContext.put(it, defaultMdc) }
                .contextWrite { MDCContext.put(it, anotherMdc) }

            StepVerifier.create(mono)
                .expectNext(2)
                .verifyComplete()
        }

        run {
            val mono = Mono.deferContextual { Mono.just(it.size()) }
                .contextWrite { MDCContext.put(it, defaultMdc, anotherMdc) }

            StepVerifier.create(mono)
                .expectNext(2)
                .verifyComplete()
        }

        run {
            val mono = Mono.deferContextual { Mono.just(it.size()) }
                .contextWrite { MDCContext.put(it, defaultMdc, null) }

            StepVerifier.create(mono)
                .expectNext(1)
                .verifyComplete()
        }

        run {
            val mono = Mono.deferContextual { Mono.just(it.size()) }
                .contextWrite { MDCContext.put(it, defaultMdc, null) }
                .contextWrite { it.put("A", "B") }

            StepVerifier.create(mono)
                .expectNext(2)
                .verifyComplete()
        }
    }

    @Test
    fun `should write within writeContext`() {
        run {
            val data = MDCContext.read()
                .mapNotNull { it["key"] }
                .contextWrite {
                    val mdc = MDCContext.getMDCOrDefault(it)
                    mdc["key"] = "example"
                    MDCContext.put(it, mdc)
                }

            StepVerifier.create(data)
                .expectNext("example")
                .verifyComplete()
        }

        run {
            val data = MDCContext.read("another")
                .mapNotNull { it["key"] }
                .contextWrite {
                    val mdc = MDCContext.getMDCOrDefault(it, "another")
                    mdc["key"] = "example"
                    MDCContext.put(it, mdc)
                }

            StepVerifier.create(data)
                .expectNext("example")
                .verifyComplete()
        }

        run {
            val data = MDCContext.read()
                .mapNotNull { it["key"] }
                .contextWrite {
                    var mdc = MDCContext.getMDCOrNull(it)
                    if (mdc == null) {
                        mdc = MDC()
                    }

                    mdc["key"] = "example"
                    MDCContext.put(it, mdc)
                }

            StepVerifier.create(data)
                .expectNext("example")
                .verifyComplete()
        }

        run {
            val data = MDCContext.read("another")
                .mapNotNull { it["key"] }
                .contextWrite {
                    var mdc = MDCContext.getMDCOrNull(it, "another")

                    if (mdc == null) {
                        mdc = MDC("another")
                        mdc!!["key"] = "example"
                    }

                    MDCContext.put(it, mdc)
                }

            StepVerifier.create(data)
                .expectNext("example")
                .verifyComplete()
        }

        run {
            val data = MDCContext.read()
                .mapNotNull { it["key"] }
                .contextWrite {
                    MDCContext.modifyContext(it) { mdc ->
                        mdc["key"] = "example"
                    }
                }

            StepVerifier.create(data)
                .expectNext("example")
                .verifyComplete()
        }

        run {
            val data = MDCContext.read("another")
                .mapNotNull { it["key"] }
                .contextWrite {
                    MDCContext.modifyContext(it, "another") { mdc ->
                        mdc["key"] = "example"
                    }
                }

            StepVerifier.create(data)
                .expectNext("example")
                .verifyComplete()
        }

        run {
            val data = MDCContext.read()
                .mapNotNull { it["key"] }
                .contextWrite { MDCContext.modifyContext(it, mapOf("key" to "example")) }

            StepVerifier.create(data)
                .expectNext("example")
                .verifyComplete()
        }

        run {
            val data = MDCContext.read("another")
                .mapNotNull { it["key"] }
                .contextWrite { MDCContext.modifyContext(it, "another", mapOf("key" to "example")) }

            StepVerifier.create(data)
                .expectNext("example")
                .verifyComplete()
        }
    }

    @Test
    fun `should give the MDC you are looking for (more MDC Context)`() {
        val defaultMdc = MDC()
        defaultMdc["mdcKey"] = "mdcValue"
        val anotherMdc = MDC(ANOTHER_CONTEXT_KEY)
        anotherMdc["mdcKey"] = "mdcValue"

        run {
            val mono: Mono<MDC> = Mono.defer { MDCContext.read() }
                .contextWrite { MDCContext.put(it, defaultMdc) }
                .contextWrite { MDCContext.put(it, anotherMdc) }

            StepVerifier.create(mono)
                .expectNext(defaultMdc)
                .verifyComplete()
        }

        run {
            val mono: Mono<MDC> = Mono.defer { MDCContext.read(ANOTHER_CONTEXT_KEY) }
                .contextWrite { MDCContext.put(it, defaultMdc) }
                .contextWrite { MDCContext.put(it, anotherMdc) }

            StepVerifier.create(mono)
                .expectNext(anotherMdc)
                .verifyComplete()
        }
    }

    @Test
    fun `should throw InvalidContextDataException if context data is invalid`() {
        val mdc = MDC(ANOTHER_CONTEXT_KEY)
        mdc["mdcKey2"] = "mdcValue2"

        run {
            val mono: Mono<MDC> = Mono.defer { MDCContext.read() }
                .contextWrite { MDCContext.put(it, mdc) }

            StepVerifier.create(mono)
                .expectNext(MDC(DEFAULT_REACTOR_CONTEXT_MDC_KEY))
                .verifyComplete()
        }

        run {
            val mono: Mono<MDC> = Mono.defer { MDCContext.read("not-exist-context-id") }
                .contextWrite { MDCContext.put(it!!, mdc) }

            StepVerifier.create(mono)
                .expectNext(MDC("not-exist-context-id"))
                .verifyComplete()
        }
    }

    companion object {
        const val ANOTHER_CONTEXT_KEY = "another-context-key"
    }
}